File: Components/WinForms/CS/GaugesMainDemo/Modules/WeatherStation.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;
using System.Xml.Serialization;
using System.Reflection;
using DevExpress.Utils;
using DevExpress.XtraGauges.Demos.net.webservicex.www;

namespace DevExpress.XtraGauges.Demos {
    public partial class WeatherStation : TutorialControl {
        BindingList<DemoWetherData> data;

        public WeatherStation() {
            InitializeComponent();
            UpdateWeather("London(Offline data)");
        }
        void simpleButton1_Click(object sender, System.EventArgs e) {
            UpdateWeather(comboBoxEdit1.EditValue.ToString());
        }
        protected void UpdateWeather(string place) {
            WeatherForecasts wfc = GetWeatherByPlaceNameSafe(place);
            if(wfc == null) return;
            InitDataSource(wfc);
            InitCharts();
            InitGauge();
        }
        protected WeatherForecasts GetWeatherByPlaceNameSafe(string place) {
            Cursor = Cursors.WaitCursor;
            WeatherForecasts wfc = null;
            try {
                using(WaitDialogForm dlg = new WaitDialogForm("Loading weather at " + place)) {
                    if(place.Contains("Offline")) {
                        XmlSerializer serializer = new XmlSerializer(typeof(WeatherForecasts));
                        wfc = (WeatherForecasts)serializer.Deserialize(Assembly.GetExecutingAssembly().GetManifestResourceStream("DevExpress.XtraGauges.Demos.Data.weather.xml"));

                    }
                    else {
                        WeatherForecast service = new WeatherForecast();
                        wfc = service.GetWeatherByPlaceName(place);
                    }
                }
            }
            catch {
                MessageBox.Show("could not retrive weather data for " + place);
                Cursor = Cursors.Arrow;
                return null;
            }
            if(wfc.Details == null) {
                MessageBox.Show("could not retrive weather data for " + place);
                Cursor = Cursors.Arrow;
                return null;
            }
            Cursor = Cursors.Arrow;
            return wfc;
        }
        protected void InitDataSource(WeatherForecasts wfc) {
            try {
                try {
                    if(data == null) data = new BindingList<DemoWetherData>();
                    else data.Clear();
                }
                catch { }
                int counter = 0;
                bool isCelsius = comboBoxEdit2.Text.StartsWith("C");

                foreach(WeatherData wd in wfc.Details) {
                    int minF, maxF, minC, maxC, f, c;
                    int.TryParse(wd.MinTemperatureF, out minF);
                    int.TryParse(wd.MaxTemperatureF, out maxF);
                    int.TryParse(wd.MinTemperatureC, out minC);
                    int.TryParse(wd.MaxTemperatureC, out maxC);
                    f = (minF + maxF) / 2; c = (minC + maxC) / 2;
                    data.Add(new DemoWetherData(isCelsius ? minC : minF, isCelsius ? maxC : maxF, c, f, counter++));
                }
            }
            catch { }
        }
        protected void InitGauge() {
            bool isCelsius = comboBoxEdit2.Text.StartsWith("C");
            if(currentTempScale.DataBindings.Count == 0) {
                currentTempScale.DataBindings.Add("Value", data, "CurrentTemperature" + (isCelsius ? "C" : "F"));
            }
        }
        protected void InitCharts() {
            if(chartControl1.DataSource != data) {
                chartControl1.DataSource = data;
                chartControl1.Series[0].ArgumentDataMember = "Day";
                chartControl1.Series[0].ValueDataMembers[0] = "MinTemperature";

                chartControl1.Series[1].ArgumentDataMember = "Day";
                chartControl1.Series[1].ValueDataMembers[0] = "MaxTemperature";
            }
            else {
                chartControl1.Invalidate();
            }

        }
    }

    public class DemoWetherData {
        int minTemperatureCore;
        int maxTemperatureCore;
        int dayCore;
        int currentTemperatureC;
        int currentTemperatureF;

        public int MinTemperature {
            get { return minTemperatureCore; }
            set { minTemperatureCore = value; }
        }
        public int MaxTemperature {
            get { return maxTemperatureCore; }
            set { maxTemperatureCore = value; }
        }
        public int Day {
            get { return dayCore; }
            set { dayCore = value; }
        }
        public int CurrentTemperatureF {
            get { return currentTemperatureC; }
            set { currentTemperatureC = value; }
        }
        public int CurrentTemperatureC {
            get { return currentTemperatureF; }
            set { currentTemperatureF = value; }
        }

        public DemoWetherData(int minT, int maxT, int currentTemperatureC, int currentTemperatureF, int day) {
            Day = day;
            MinTemperature = minT;
            MaxTemperature = maxT;
            CurrentTemperatureC = currentTemperatureC;
            CurrentTemperatureF = currentTemperatureF;
        }
    }
}

File: Components/WinForms/VB/GaugesMainDemo/Modules/WeatherStation.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Reflection
Imports System.Windows.Forms
Imports System.Xml.Serialization
Imports DevExpress.Utils

Namespace DevExpress.XtraGauges.Demos
    Partial Public Class WeatherStation
        Inherits TutorialControl

        Private data As BindingList(Of DemoWetherData)

        Public Sub New()
            InitializeComponent()
            UpdateWeather("London(Offline data)")
        End Sub
        Sub simpleButton1_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles simpleButton1.Click
            UpdateWeather(comboBoxEdit1.EditValue.ToString())
        End Sub
        Protected Sub UpdateWeather(ByVal place As String)
            Dim wfc As net.webservicex.www.WeatherForecasts = GetWeatherByPlaceNameSafe(place)
            If wfc Is Nothing Then
                Return
            End If
            InitDataSource(wfc)
            InitCharts()
            InitGauge()
        End Sub
        Protected Sub InitDataSource(ByVal wfc As net.webservicex.www.WeatherForecasts)
            Try
                Try
                    If data Is Nothing Then
                        data = New BindingList(Of DemoWetherData)()
                    Else
                        data.Clear()
                    End If
                Catch
                End Try
                Dim counter As Integer = 0
                For Each wd As net.webservicex.www.WeatherData In wfc.Details
                    Dim minT As String
                    If comboBoxEdit2.Text.StartsWith("C") Then
                        minT = wd.MinTemperatureC
                    Else
                        minT = wd.MinTemperatureF
                    End If
                    Dim maxT As String
                    If comboBoxEdit2.Text.StartsWith("C") Then
                        maxT = wd.MaxTemperatureC
                    Else
                        maxT = wd.MaxTemperatureF
                    End If
                    Dim min As Integer = Integer.Parse(minT)
                    Dim max As Integer = Integer.Parse(maxT)

                    data.Add(New DemoWetherData(min, max, (min + max) \ 2, counter))
                    counter += 1
                Next wd
            Catch
            End Try
        End Sub
        Protected Sub InitGauge()
            If linearScaleComponent1.DataBindings.Count = 0 Then
                linearScaleComponent1.DataBindings.Add("Value", data, "CurrentTemperature")
            End If
        End Sub
        Protected Sub InitCharts()
            If chartControl1.DataSource IsNot data Then
                chartControl1.DataSource = data
                chartControl1.Series(0).ArgumentDataMember = "Day"
                chartControl1.Series(0).ValueDataMembers(0) = "MinTemperature"

                chartControl1.Series(1).ArgumentDataMember = "Day"
                chartControl1.Series(1).ValueDataMembers(0) = "MaxTemperature"
            Else
                chartControl1.Invalidate()
            End If

        End Sub
        Protected Function GetWeatherByPlaceNameSafe(ByVal place As String) As net.webservicex.www.WeatherForecasts
            Cursor = Cursors.WaitCursor
            Dim wfc As net.webservicex.www.WeatherForecasts = Nothing
            Try
                Using dlg As WaitDialogForm = New WaitDialogForm("Loading weather at " & place)
                    If place.Contains("Offline") Then
                        Dim serializer As XmlSerializer = New XmlSerializer(GetType(net.webservicex.www.WeatherForecasts))
                        wfc = CType(serializer.Deserialize(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("weather.xml")), net.webservicex.www.WeatherForecasts)
                    Else
                        Dim service As net.webservicex.www.WeatherForecast = New net.webservicex.www.WeatherForecast()
                        wfc = service.GetWeatherByPlaceName(place)
                    End If
                End Using
            Catch
                MessageBox.Show("could not retrive weather data for " & place)
                Cursor = Cursors.Arrow
                Return Nothing
            End Try
            If wfc.Details Is Nothing Then
                MessageBox.Show("could not retrive weather data for " & place)
                Cursor = Cursors.Arrow
                Return Nothing
            End If
            Cursor = Cursors.Arrow
            Return wfc
        End Function
    End Class

    Public Class DemoWetherData
        Private minTemperatureCore As Integer
        Private maxTemperatureCore As Integer
        Private dayCore As Integer
        Private currentTemperature_Renamed As Integer

        Public Property MinTemperature() As Integer
            Get
                Return minTemperatureCore
            End Get
            Set(ByVal value As Integer)
                minTemperatureCore = value
            End Set
        End Property
        Public Property MaxTemperature() As Integer
            Get
                Return maxTemperatureCore
            End Get
            Set(ByVal value As Integer)
                maxTemperatureCore = value
            End Set
        End Property
        Public Property Day() As Integer
            Get
                Return dayCore
            End Get
            Set(ByVal value As Integer)
                dayCore = value
            End Set
        End Property
        Public Property CurrentTemperature() As Integer
            Get
                Return currentTemperature_Renamed
            End Get
            Set(ByVal value As Integer)
                currentTemperature_Renamed = value
            End Set
        End Property

        Public Sub New(ByVal minT As Integer, ByVal maxT As Integer, ByVal currentTemperature As Integer, ByVal day As Integer)
            Me.Day = day
            MinTemperature = minT
            MaxTemperature = maxT
            Me.CurrentTemperature = currentTemperature
        End Sub
    End Class
End Namespace
